### PR TITLE
Non-blocking lock for controller manager

### PR DESCRIPTION
**Pull request by qiujian16**
still working on integration test.

### CODE REVIEW

File: pkg/db/advisory_locks.go
Comment (markturansky) [inline pkg/db/advisory_locks.go:28]: I think _nonblocking_ is technically more correct than unblocking.

**Comment by markturansky** [comment]
This is a fantastic addition. The current test for blocking locks looks good.  we need 1 for non-blocking locks.

File: pkg/db/advisory_locks.go
Comment (qiujian16) [reply pkg/db/advisory_locks.go]: change the name to nonblocking

**Comment by qiujian16** [comment]
> This is a fantastic addition. The current test for blocking locks looks good. we need 1 for non-blocking locks.

test is added.

**Comment by qiujian16** [comment]
this should hold until rebase after #3 is merged

**Comment by qiujian16** [comment]
cc @skeeey

**Comment by markturansky** [comment]
This looks good. The tests are perfect. We'll wait until #3 is merged.

File: pkg/controllers/framework.go
Comment (markturansky) [inline pkg/controllers/framework.go:96]: this comment can be removed.

Comment (markturansky) [reply pkg/controllers/framework.go:96]: parallel execution happens by having more than 1 replica of our deployed pod.

Comment (qiujian16) [reply pkg/controllers/framework.go]: removed

**Comment by markturansky** [comment]
@qiujian16 this needs rebase

File: test/integration/controller_test.go
Comment (markturansky) [inline test/integration/controller_test.go:80]: great test. 3 controller and 2 events to process.

**Review by markturansky — APPROVED** [review]
LGTM

**Comment by markturansky** [comment]
@qiujian16 did you run the integration tests?

I get this (but I'm also on a different dev branch I'm rebasing):

```
fatal error: concurrent map writes

goroutine 344 [running]:
github.com/openshift-online/rh-trex/pkg/db.(*AdvisoryLockFactory).Unlock(0xc000447710, {0x14dd330?, 0xc0008e29f0?}, {0xc0002bc090, 0x24})
	/home/mturansk/projects/src/github.com/openshift-online/rh-trex/pkg/db/advisory_locks.go:137 +0x3e5
github.com/openshift-online/rh-trex/pkg/services.(*sqlDinosaurService).Replace(0xc0001f5fb0, {0x14dd330, 0xc0008e29f0}, 0xc0002921c0)
	/home/mturansk/projects/src/github.com/openshift-online/rh-trex/pkg/services/dinosaurs.go:112 +0x1c5
github.com/openshift-online/rh-trex/pkg/handlers.dinosaurHandler.Patch.func1()
	/home/mturansk/projects/src/github.com/openshift-online/rh-trex/pkg/handlers/dinosaur.go:63 +0x137
```

**Comment by qiujian16** [comment]
hrm no, but it seems like a race when writing the map for lock. Will take a look.

**Pull request merged**
